chainstate/src/detail/chainstateref/consistency_checker.rs

```diff
@@ -0,0 +1,268 @@
+
+        // TODO: add consistency checks for other maps in the chainstate db.
+        // https://github.com/mintlayer/mintlayer-core/issues/1710
+
+        Ok(())
+    }
+
+    /// Check the block map vs block index map consistency.
+    fn check_block_index_consistency(&self) -> Result<(), chainstate_storage::Error> {
+        // Loop over block_map_keys and block_index_map simultaneously via merge_join_by, looking
+        // for ids that are present in one of them and missing in the other.
+        for merged in self
+            .block_map_keys
+            .iter()
+            .merge_join_by(self.block_index_map.iter(), |id1, (id2, _)| {
+                Ord::cmp(id1, id2)
+            })
+        {
+            let (block_id, block_index) = match merged {
+                EitherOrBoth::Left(block_id) => {
+                    // The block object is present, the index object is not.
+                    panic!("{PANIC_MSG}: block index data missing for block {block_id}");
+                }
+                EitherOrBoth::Right((block_id, block_index)) => {
+                    // The block index object is present, the block object is not;
+                    // The persistence flag must be unset and the status must not be "ok".
+                    assert!(
+                        !block_index.is_persisted(),
+                        "{PANIC_MSG}: block {block_id} must not be persisted"
+                    );
+                    assert!(
+                        !block_index.status().is_ok(),
+                        "{PANIC_MSG}: block {block_id} must not be ok"
+                    );
+
+                    (block_id, block_index)
+                }
+                EitherOrBoth::Both(_, (block_id, block_index)) => {
+                    // Both the block and block index objects are present.
+
+                    // The persistence flag must be set.
+                    assert!(
+                        block_index.is_persisted(),
+                        "{PANIC_MSG}: block {block_id} must be persisted"
+                    );
+
+                    (block_id, block_index)
+                }
+            };
+
+            // Check that the id stored in the block index matches the supposed id of the block.
+            let block_id_in_block_index = block_index.block_id();
+            assert_eq!(
+                block_id,
+                block_id_in_block_index,
+                "{PANIC_MSG}: block id from BlockIndex {block_id_in_block_index} doesn't match {block_id}"
+            );
+
+            // If the block is persisted, calculate its id and check that it matches the id
+            // that was used as the key. Also compare the block header stored in the index vs the one
+            // in the block itself.
+            // Note that this check is relatively heavy, because it has to load the whole block from the db,
+            // so we do it only in release builds.
+            #[cfg(not(debug_assertions))]
```

Comment:
Is this really necessary?

Enabling debug_assertions doesn't necessarily mean it's a release build.

Reply:
> Is this really necessary?

With this check disabled in debug, `cargo nextest run --workspace` takes 1m 24s on my laptop, with it enabled - 3m 31s. So yeah, it's noticeable. If I disable all the checks in the slowest test (syncing_tests::get_mainchain_headers_by_locator), like I already do for another test, the whole run now takes 1m 53s.
Probably, this is a better approach, so I'll leave it like this. Though the extra 30s are annoying too.

Another thing is that this particular check is not that useful, so probably just removing it might be better at this moment. But in the future we'll implement more checks and some of them will require loading whole blocks from the db, which is what makes it slow. So we'll return to this problem.